### ISSUE TITLE
Access to generic instance variable through class is ambiguous

### ISSUE THREAD

**Issue by nanne-aben** [microsoft/pyright#7089]
**Describe the bug**
As of v1.1.348, I'm getting the following error:

```
error: Access to generic instance variable through class is ambiguous (reportGeneralTypeIssues)
```

**Code or Screenshots**
Minimal example:
```python
from typing import Generic, Type, TypeVar

T = TypeVar("T")


class Bar:
    pass


class Foo(Generic[T]):
    bar: Type[T]


Foo[Bar].bar
```

I don't understand why this is ambiguous though. I'd think that it is pretty clear that the type of `Foo[Bar].bar` is `Type[Bar]`. Happy to learn if that's not the case though!

In context, the error can be seen in the ci/cd logs of https://github.com/kaiko-ai/typedspark/pull/281, if you're interested.

**VS Code extension or command-line**
From the command line, v1.1.348. Previous versions did not raise this error.

Thanks in advance!

**Comment by erictraut** [comment]
At runtime, all type information is "erased". That means the class `Foo` in your example is the same regardless of whether you're treating it as `Foo[str]` or `Foo[int]`. And that means that if you assign an `int` value via the statement `Foo[int].bar = 1` and then later access the value `Foo[str].bar`, it is not a `str` as you might expect.

You can read more about this in the [official typing spec](https://typing.readthedocs.io/en/latest/spec/generics.html#instantiating-generic-classes-and-type-erasure). Pyright was previously missing this check, which is mandated by the spec.

**Comment by lorenzo-w** [comment]
For me this also raises when accessing the variables of a generic class through `cls` inside a classmethod. Shouldn't that be correct usage?

**Comment by erictraut** [comment]
> Shouldn't that be correct usage?

No, that's a problem for the same reason I explained above. The `cls` refers to a singleton class object, and its generic type is erased at runtime.

**Comment by nanne-aben** [comment]
@erictraut thanks for your answer, it helped me solve the problem!

**Comment by last-partizan** [comment]
I'm stumbled upon this when upgrading as well.

Consider following example:

```python
from typing import Callable, Generic, Sequence, TypeVar

T = TypeVar("T")

class Admin(Generic[T]):
    list_display: Sequence[str | Callable[[T], str]] = ("__str__", )

class CommonAdmin(Admin):
    list_display = ("title", )

class BookAdmin(CommonAdmin):
    list_display = (*CommonAdmin.list_display, "author")  # - error: Access to generic instance variable through class is ambiguous (reportGeneralTypeIssues)
```

Here's I'm not accessing generic variable directly, i'm accessing sequence including generic variable, and that's should be fine.

@erictraut what do you think?

I'm fine with adding few `pyright: ignore`, but maybe it can be fixed, or there's another way to write this without triggering this error?

**Comment by erictraut** [comment]
> Here's I'm not accessing generic variable directly

But you are accessing a generic variable. It's a variable that has a declared type of `Sequence[str | Callable[[T], str]]`, which is generic. It looks like the variable `list_display` is always used as a class variable, never as an instance variable. If it's truly meant to be a class variable, you should probably declare it as such.

```python
class Admin(Generic[T]):
    list_display: ClassVar[Sequence[str | Callable[[T], str]]] = ("__str__",)
```

But this (correctly) results in another error because a `ClassVar` cannot be generic for the same reason that it's unsafe to access a generic instance variable through a class object.

Perhaps what you really intend is this?

```python
class Admin(Generic[T]):
    list_display: ClassVar[Sequence[str | Callable[[Any], str]]] = ("__str__",)
```

**Comment by last-partizan** [comment]
Yes, it will work. But we're losing type information about T in this case, so i'm not gonna commit this change into django-types.

But, probably it's best to override type as `list_display: Sequence[str]` in my `CommonAdmin`.

Thanks!

**Comment by bricker** [comment]
@erictraut I could use some help understanding this. In the case of sub-classing the Generic class and explicitly giving the type `T`, what runtime problem could this cause?

Example:

```python
from typing import Generic, TypeVar

T = TypeVar("T")

class Base(Generic[T]):         some_attr: T
class StringBase(Base[str]):    some_attr = "foo"
class IntBase(Base[int]):       some_attr = 0

assert StringBase.some_attr == "foo" # `Access to generic instance variable through class is ambiguous`
assert IntBase.some_attr    == 0     # `Access to generic instance variable through class is ambiguous`
```

This still gives the type-checker error, but to me this is not ambiguous. At runtime, it is equivalent to:

```python
class Base:             pass
class StringBase(Base): some_attr = "foo"
class IntBase(Base):    some_attr = 0

assert StringBase.some_attr == "foo"
assert IntBase.some_attr    == 0
```

What am I missing?

**Comment by erictraut** [comment]
@bricker, I agree it's not ambiguous in the code that you posted. This is different from the OP's case. In your code sample, I agree that pyright should not emit an error. Could you open a new issue to track this?

**Comment by olafurnielsen** [comment]
@bricker, @erictraut – Was another issue created for the generic base class use case where the type is explicit in subclasses? Don't seem to find it and would like to subscribe to updates on that.
